server/src/main/java/org/elasticsearch/action/support/DefaultShardOperationFailedException.java

```diff
@@ -22,17 +22,31 @@
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.action.ShardOperationFailedException;
+import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.rest.RestStatus;
 
 import java.io.IOException;
 
 import static org.elasticsearch.ExceptionsHelper.detailedMessage;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
 
 public class DefaultShardOperationFailedException implements ShardOperationFailedException {
 
+    private static final ConstructingObjectParser<DefaultShardOperationFailedException, Void> PARSER = new ConstructingObjectParser<>(
+        "failures", true, arg -> new DefaultShardOperationFailedException((String) arg[0], (int) arg[1], (Throwable) arg[2]));
+
+    static {
+        PARSER.declareString(constructorArg(), new ParseField(Fields.INDEX));
+        PARSER.declareInt(constructorArg(), new ParseField(Fields.SHARD_ID));
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> ElasticsearchException.fromXContent(p), new ParseField(Fields.REASON));
```

Comment:
I take it that you deliberately skipped status, I wonder if we should also skip index and shard, given that we will always have an ElasticsearchException back we should rather use the `DefaultShardOperationFailedException(ElasticsearchException)` constructor all the time, then index, shard and status will be extracted from such exception.